### PR TITLE
Cancellation token sync

### LINKED ISSUE #230: Pass CancellationToken To RunAsync and ReadAsync
## Use case(s)

Some processes are long-running tasks, and the ability to `kill` these processes might be a good thing. I have a process that starts playing an audio file and I'd like to stop the audio with a cancellationToken.

## Description

```
            var audio = await SimpleExec.Command.ReadAsync(
                "afplay",
                string.Join(" ", arguments),
                noEcho: true,
                cancellationToken: cancellationToken
            );
```

The issue is that the cancellation request needs to call `process.Kill` or else the process continues to run in the background.

## Alternatives

Handroll starting a process.

## Additional context

I'll try and make an attempt myself at implementing the change, but my `async` foo is mediocre at best.

### PR DESCRIPTION
**Pull request by danielmarbach**
Closes #230

As discussed in https://github.com/adamralph/simple-exec/pull/238#issuecomment-731807613

### CODE REVIEW

**Comment by danielmarbach** [comment]
@adamralph nuget push step seems to fail on PRs

File: SimpleExec/ProcessExtensions.cs [hunk 1]
```diff
@@ -1,16 +1,54 @@
...
-            process.EchoAndStart(noEcho, echoPrefix);
-            process.WaitForExit();
+            using (var resetEvent = new ManualResetEventSlim(false))
+            {
+                StringBuilder stringBuilder = null;
+                process.Exited += (s, e) => resetEvent.SafeSet();
+                if(readStandardOutput)
+                {
+                    stringBuilder = new StringBuilder();
+                    process.OutputDataReceived += (s, e) => stringBuilder.AppendLine(e.Data);
```
Comment (danielmarbach) [inline SimpleExec/ProcessExtensions.cs:20]: Once we have settled on how you want this code to look like we could apply the same to the async path if desired for consistency

**Review by adamralph — COMMENTED** [review]
Generally looks good. I just have one concern.

BTW — this is a naive question, but what is the problem with simply wrapping both `RunAsync` and `ReadAsync` with `.GetAwaiter().GetResult()` to provide sync convenience methods? That's actually what I'm currently doing in Bullseye, and I guess I'm just getting lucky and avoiding any problems so far.

TBH, and this may just be a brain fart, but given the naturally async nature of what SimpleExec does, I'm even considering dropping the sync methods entirely (and bumping the major version).

File: SimpleExec/ProcessExtensions.cs [hunk 2]
```diff
@@ -1,16 +1,54 @@
...
+                }
+                process.EnableRaisingEvents = true;
+                process.EchoAndStart(noEcho, echoPrefix);
+                if (readStandardOutput)
+                {
+                    process.BeginOutputReadLine();
```
[the remark this answers is not in the document]
Comment (adamralph) [reply SimpleExec/ProcessExtensions.cs:26]: I think the test failures in the [windows build](https://github.com/adamralph/simple-exec/pull/242/checks?check_run_id=1461191585) may validate this concern. It looks like, in some cases, stdout is winning the race and this line is not capturing the expected output.

[the remark this answers is not in the document]
Comment (danielmarbach) [reply SimpleExec/ProcessExtensions.cs:26]: I think I can fix this race but the more important question is if you stick to having sync version

**Comment by danielmarbach** [comment]
Don't know where this tool is mostly used. You are the best to judge. Maybe for simplicity for the users it is valid to have sync overloads. If you do sync over async is dangerous and wasteful due to needing more threads than a pure sync version. That being said it might be an acceptable tradeoff for this tool. I don't knowAm 27.11.2020 17:07 schrieb Adam Ralph <notifications@github.com>:

@adamralph commented on this pull request.



Generally looks good. I just have one concern.

BTW — this is a naive question, but what is the problem with simply wrapping both RunAsync and ReadAsync with .GetAwaiter().GetResult() to provide sync convenience methods? That's actually what I'm currently doing in Bullseye, and I guess I'm just getting lucky and avoiding any problems so far.

TBH, and this may just be a brain fart, but given the naturally async nature of what SimpleExec does, I'm even considering dropping the sync methods entirely (and bumping the major version).



In SimpleExec/ProcessExtensions.cs:

> -            process.EchoAndStart(noEcho, echoPrefix);

-            process.WaitForExit();

+            using (var resetEvent = new ManualResetEventSlim(false))

+            {

+                StringBuilder stringBuilder = null;

+                process.Exited += (s, e) => resetEvent.SafeSet();

+                if(readStandardOutput)

+                {

+                    stringBuilder = new StringBuilder();

+                    process.OutputDataReceived += (s, e) => stringBuilder.AppendLine(e.Data);

+                }

+                process.EnableRaisingEvents = true;

+                process.EchoAndStart(noEcho, echoPrefix);

+                if (readStandardOutput)

+                {

+                    process.BeginOutputReadLine();



I'm a little nervous about this bit. IIRC the race between the first characters printed to stdout and this line can be significant. In early iterations of SimpleExec I played with this, but I ended up settling on process.StandardOutput.ReadToEndAsync() since that operates on the entire stream. However, I guess that is where you ran into problems with deadlocks, calling that after waiting for the process to end?

**Comment by adamralph** [comment]
> If you do sync over async [it] is ... wasteful

I think I understand this part, but I don't think it's a problem for SimpleExec. It's not something used on a hot path to spin up thousands of processes, and the overhead of the external process invocation itself likely outweighs this overhead by a large factor.

> If you do sync over async [it] is dangerous

This is the part I've never really understood. Why is it dangerous?

**Comment by danielmarbach** [comment]
Old but still relevant https://devblogs.microsoft.com/pfxteam/should-i-expose-synchronous-wrappers-for-asynchronous-methods/ If you only provide async methods then synchronisation is a duty of the caller of the Api and that caller has the necessary context whether sync over async is OK or not. If the libs does it for them and they use it in a "contextual" environment they are likely to run into a deadlock. Of course it is valid to say this lib should never be used in such an environment and thus sync over async might be OK because you cannot afford to maintain two code paths. If these tradeoffs can be made fine. Otherwise this would show how to achieve a proper sync version

**Comment by adamralph** [comment]
Never mind, I found [this article](https://docs.microsoft.com/en-us/archive/msdn-magazine/2015/july/async-programming-brownfield-async-development#the-blocking-hack) and I think I understand it now.

I've also remembered that we have another option here: don't both supporting `CancellationToken` for the sync methods. No-one has requested this, so I wonder if that might be the best option.

**Comment by adamralph** [comment]
Our comments overlapped. 🙂

**Comment by adamralph** [comment]
@danielmarbach I'm leaning towards dropping this and only supporting `CancellationToken` in the async methods. I appreciate your efforts on this though, and the _vast_ simplification you brought to the [async implementation](https://github.com/adamralph/simple-exec/pull/231).

Another result of your input is that I've realised that `Read` is liable to deadlocks, since it's just doing a "dumb" sync-over-aync with `Task.WaitAll()`. I considered removing it, but given that it's fine for most scenarios, and probably 99.9% of cases where SimpleExec is used, I decided to keep it and mitigate the problem with a [remark](https://github.com/adamralph/simple-exec/pull/249).

**Comment by danielmarbach** [comment]
Perfect. Thanks for the update. I'll close the PR

**Pull request closed without merging**